File: Alignment/MillePedeAlignmentAlgorithm/scripts/mps_parse_pedechi2hist.pl
#!/usr/bin/env perl

# Author: Joerg Behr
#
# This script reads the histogram file produced by Pede and it extracts the plot showing the average chi2/ndf per Mille binary number.
# After reading the MPS library, for which the file name has to be provided as first argument, an output file called chi2pedehis.txt is produced
# where the first column corresponds to the associated name, the second column corresponds to the Mille binary number, and the last column
# is equal to <chi2/ndf>. As second argument this scripts expects the file name of the Pede histogram file -- usually millepede.his. The last argument
# represents the location of the Python config which was used by CMSSW.
#
# Standalone usage:
# mps_parse_pedechi2hist.pl mps.db millepede.his $RUNDIR/alignment_merge.py
#
# Use createChi2ndfplot.C to plot the output of this script.

use strict;

my $db = $ARGV[0];
my $pedehis = $ARGV[1];
my $pyconfig = $ARGV[2];


if(!defined $ARGV[0] || !defined $ARGV[1] || !defined $ARGV[2])
  {
    print "The location of the mps.db and Pede histogram file, and the Python config for CMSSW have to be specified.\n";
    exit 0;
  }

unless(-e "$db")
  {
    print "Could not find mps.db file: $db\n";
    exit 0;
  }

unless(-e "$pedehis")
  {
    print "Could not find pedehis file: $pedehis\n";
    exit 0;
  }

unless(-e "$pyconfig")
  {
    print "Could not find Python config file: $pyconfig\n";
    exit 0;
  }


open IN, "< $db" || die "error while opening mps.db file: $!";
my @adb = <IN>;
close IN;

my @ids;
my @names;
my $lastid = 0;

#read the file and apply a chain of filters.
foreach (@adb)
  {
    next unless(/job/);

    if(my $n = () = /\:/g)
      {
        next if ($n != 12);
      }
    my @row = split /:/;
    my $nrow = @row;
    if($nrow != 13)
      {
        print "There seems to be a problem reading the file mps.db. The number of fields per row is $nrow but it should be equal to 13.\n";
        exit 0;
      }
    next if(!defined $row[0] || !defined $row[1]);
    next if($row[0] =~ /\D/g);
    my $jobid = $row[1];
    next if($jobid =~ /jobm/);
    $jobid =~ s/job//;
    next if($jobid =~ /\D/g);
    next if($jobid != $row[0]);
    my $name = pop @row;

    chomp $name;
    next if(!defined $name);
    next if($name eq "");
    push @names, $name;
    push @ids, $jobid;
    $lastid = $jobid if($jobid > $lastid);
  }

if(@names == 0)
  {
    print "Names were not associated to Mille binaries (Use mps_setup.pl -N blabla ... for that purpose). Chi2ndf plot is not produced. exit.\n";
    exit(-1);
  }
else
  {
    #now try to guess the used binaries. extract the file number of those
    open INPY, "< $pyconfig" || die "error while opening alignment_merge.py: $!\n";
    my @pycontent = <INPY>;
    close INPY;
    chomp @pycontent;

    my @usedbinaries;
    foreach (@pycontent)
      {
        next unless(/milleBinary/i);
        if(my ($i) = /milleBinary(.+?)\.dat/i)
          {
            push @usedbinaries, int scalar $i;
          }
      }
    #read pede hists
    open INPEDEHIS, "< $pedehis" || die "error while opening pedehis file: $!\n";
    my @ph = <INPEDEHIS>;
    close INPEDEHIS;
    chomp @ph;
    my $foundchi2start = 0;
    my @hisdata;
    for(my $i =0; $i < @ph; $i++)
      {
        my $line = $ph[$i];
        $foundchi2start = 1 if ($line =~ /final \<Chi\^2\/Ndf\> from accepted local fits vs file number/i);
        if($foundchi2start)
          {
            last if($line =~ /end of xy\-data/i);
          }
        next unless($line =~ /\d/);
        next if($line =~ /[a-z]/i);
        if($foundchi2start)
          {
            my @tmp = split " ", $line;
            $hisdata[int scalar $tmp[0] - 1] = $tmp[1];
          }
      }
    if(scalar @hisdata != scalar @usedbinaries)
      {
        print "The number of used binaries is " . scalar @usedbinaries . " whereas in contrast, however, the <chi2/ndf> histogram in Pede has " . scalar @hisdata . " bins (Pede version >= rev92 might help if #bins < #binaries). exit.\n";
        exit 0;
      }

    my $outname = "chi2pedehis.txt";
    unlink "$outname" if(-e "$outname");
    open OUTPUT, "> $outname" || die "error while opening file for writing: $!\n";

    for(my $i = 1; $i <= @usedbinaries; $i++)
     {
       my $found = 0;
       my $name;
       my $index = -1;
       foreach my $j (@ids)
         {
           if($usedbinaries[$i-1] == $j)
             {
               $found = 1;
               $name  = $names[$j-1];
               $index = $j;
               last;
             }
         }
       if(!$found)
         {
           print "Problem parsing the mps.db file. Could not find job $i. Maybe only a fraction of binaries was named?\n";
           exit 0;
         }
       else
         {
           my $tmp2 =  $ids[$index-1];
           if(defined $hisdata[$i-1] && defined $name && defined $tmp2)
             {
               print OUTPUT "$name $tmp2 " . $hisdata[$i-1] . "\n";
             }
         }
     }
    close OUTPUT;
  }

File: Alignment/MillePedeAlignmentAlgorithm/scripts/mps_parse_pedechi2hist.py
#!/usr/bin/env python3

# Original author: Joerg Behr
# Translation from Perl to Python: Gregor Mittag
#
# This script reads the histogram file produced by Pede and it extracts the plot
# showing the average chi2/ndf per Mille binary number.  After reading the MPS
# database, for which the file name has to be provided, an output file called
# chi2pedehis.txt is produced where the first column corresponds to the
# associated name, the second column corresponds to the Mille binary number, and
# the last column is equal to <chi2/ndf>. As further argument this scripts
# expects the file name of the Pede histogram file -- usually millepede.his. The
# last required argument represents the location of the Python config which was
# used by CMSSW.
#
# Use createChi2ndfplot.C to plot the output of this script.

from __future__ import print_function
import os
import sys
import re
import argparse

import Alignment.MillePedeAlignmentAlgorithm.mpslib.tools as mps_tools
import Alignment.MillePedeAlignmentAlgorithm.mpslib.Mpslibclass as mpslib


################################################################################
def main(argv = None):
    """Main routine of the script.

    Arguments:
    - `argv`: arguments passed to the main routine
    """

    if argv == None:
        argv = sys.argv[1:]

    parser = argparse.ArgumentParser(description="Analysis pede histogram file")
    parser.add_argument("-d", "--mps-db", dest="mps_db", required=True,
                        metavar="PATH", help="MPS database file ('mps.db')")
    parser.add_argument("--his", dest="his_file", required=True,
                        metavar="PATH", help="pede histogram file")
    parser.add_argument("-c", "--cfg", dest="cfg", metavar="PATH", required=True,
                        help="python configuration file of pede job")
    parser.add_argument("-b", "--no-binary-check", dest="no_binary_check",
                        default=False, action="store_true",
                        help=("skip check for existing binaries "
                              "(possibly needed if used interactively)"))
    args = parser.parse_args(argv)


    for input_file in (args.mps_db, args.his_file, args.cfg):
        if not os.path.exists(input_file):
            print("Could not find input file:", input_file)
            sys.exit(1)

    ids, names = get_all_ids_names(args.mps_db)
    used_binaries = get_used_binaries(args.cfg, args.no_binary_check)
    his_data = get_his_data(args.his_file)

    if len(his_data) != len(used_binaries):
        print("The number of used binaries is", len(used_binaries), end=' ')
        print("whereas in contrast, however, the <chi2/ndf> histogram in Pede has", end=' ')
        print(len(his_data), "bins (Pede version >= rev92 might help if #bins < #binaries).", end=' ')
        print("Exiting.")
        sys.exit(1)

    with open("chi2pedehis.txt", "w") as f:
        for i, b in enumerate(used_binaries):
            index = ids.index(b)
            name = names[index]
            f.write(" ".join([name, "{:03d}".format(b), his_data[i]])+"\n")


################################################################################
def get_all_ids_names(mps_db):
    """Returns two lists containing the mille job IDs and the associated names.
    
    Arguments:
    - `mps_db`: path to the MPS database file
    """

    lib = mpslib.jobdatabase()
    lib.read_db(mps_db)

    ids = lib.JOBNUMBER[:lib.nJobs]
    names = lib.JOBSP3[:lib.nJobs]

    return ids, names


def get_used_binaries(cfg, no_binary_check):
    """Returns list of used binary IDs.
    
    Arguments:
    - `cfg`: python config used to run the pede job
    - `no_binary_check`: if 'True' a check for file existence is skipped
    """

    cms_process = mps_tools.get_process_object(cfg)

    binaries = cms_process.AlignmentProducer.algoConfig.mergeBinaryFiles
    if no_binary_check:
        used_binaries = binaries
    else:
        # following check works only if 'args.cfg' was run from the same directory:
        used_binaries = [b for b in binaries
                         if os.path.exists(os.path.join(os.path.dirname(cfg), b))]

    used_binaries = [int(re.sub(r"milleBinary(\d+)\.dat", r"\1", b))
                     for b in used_binaries]

    return used_binaries


def get_his_data(his_file):
    """Parse the pede histogram file.
    
    Arguments:
    - `his_file`: pede histogram file
    """
    
    his_data = []
    with open(his_file, "r") as his:
        found_chi2_start = False;
        
        for line in his:
            if r"final <Chi^2/Ndf> from accepted local fits vs file number" in line:
                found_chi2_start = True
            if not found_chi2_start:
                continue
            else:
                if r"end of xy-data" in line: break
                if not re.search("\d", line): continue
                if re.search(r"[a-z]", line): continue
                splitted = line.split()
                his_data.append(splitted[-1])

    return his_data

    
################################################################################
if __name__ == "__main__":
    main()
